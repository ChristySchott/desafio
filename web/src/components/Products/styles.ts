import styled from 'styled-components';

export const Container = styled.div`
  width: 90%;
  margin: 0 auto;

  h2 {
    color: #c61a28;
    font-weight: 500;
    text-transform: capitalize;
  }
`;

export const Menu = styled.div`
  display: flex;
  align-items: center;
  justify-content: space-between;
  margin-top: 8px;

  @media (min-width: 768px) {
    border-top: 1px solid #ddd;
    border-bottom: 1px solid #ddd;
    padding: 8px 0;
  }
`;

export const OrderBy = styled.div`
  display: flex;
  align-items: center;
  justify-content: space-between;
  width: 100%;

  span {
    color: #888;
    font-weight: 600;
    text-transform: uppercase;
    font-size: 1.2rem;
    margin-right: 10px;
  }

  @media (min-width: 768px) {
    justify-content: flex-end;
  }
`;

export const View = styled.div`
  display: none;
  padding-left: 2px;

  svg + svg {
    margin-left: 5px;
  }

  @media (min-width: 768px) {
    display: flex;
    align-items: center;
    justify-content: space-between;
  }
`;

export const Items = styled.ul`
  display: flex;
  align-items: flex-start;
  flex-wrap: wrap;
  justify-content: space-between;

  list-style: none;
`;
