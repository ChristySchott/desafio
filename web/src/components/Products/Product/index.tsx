/* eslint-disable import/no-dynamic-require */
/* eslint-disable global-require */
import React from 'react';

import formatValue from 'utils/formatValue';
import { ProductInterface, useCart } from 'hooks/cart';
import { Container, Image, Infos } from './styles';

interface ProductProps {
  imageUrl: string;
  alt: string;
  name: string;
  offer?: number;
  price: number;
  item: ProductInterface;
}

const Product: React.FC<ProductProps> = ({ alt, name, offer, price, item }) => {
  const { addToCart } = useCart();

  function handleAddToCart(product: ProductInterface): void {
    addToCart(product);
  }

  return (
    <Container>
      <a href="/">
        <Image>
          <img src={require('assets/products/shoes-6.jpg')} alt={alt} />
        </Image>
      </a>

      <Infos>
        <span>{name}</span>
        {!!offer && <span>{formatValue(offer)}</span>}
        <strong>{formatValue(price)}</strong>
      </Infos>

      <button type="button" onClick={() => handleAddToCart(item)}>
        Comprar
      </button>
    </Container>
  );
};

export default Product;
