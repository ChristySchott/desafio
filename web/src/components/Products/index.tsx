/* eslint-disable react/no-array-index-key */
/* eslint-disable react-hooks/exhaustive-deps */
import React, { useEffect, useState } from 'react';
import { BsGrid3X3GapFill } from 'react-icons/bs';
import { RiLayout2Fill } from 'react-icons/ri';

import shirt from 'assets/products/shirt-1.jpg';
import api from 'services/api';
import { useCategory } from 'hooks/category';
import { useFilter } from 'hooks/filter';
import { useColorFilter } from 'hooks/color';
import TagFilter from 'components/TagFilter';
import { Container, View, Menu, OrderBy, Items } from './styles';
import Product from './Product';

import Select from '../Select';
import Paginator from '../Paginator';

interface OptionTypes {
  value: string;
  label: string;
}

const orderOptions: OptionTypes[] = [
  { value: 'preço', label: 'Preço' },
  { value: 'data', label: 'Data' },
];

const Products: React.FC = () => {
  const { category } = useCategory();
  const { filter, setFilter } = useFilter();
  const { colorToFilter, setColorToFilter } = useColorFilter();
  const [products, setProducts] = useState([]);

  let filters = '';
  let filterByType = '';
  useEffect(() => {
    if (filter) {
      filters = `?&filter.type=${filter}`;
      filterByType = `&filter.type=${filter}`;
    }

    if (colorToFilter) {
      filters = `?${filterByType}&filter.color=${colorToFilter}`;
    }
  }, [filter, colorToFilter]);

  useEffect(() => {
    async function loadProducts(): Promise<void> {
      const response = await api.get(`${category}${filters}`);
      setProducts(response.data);
    }

    loadProducts();
  }, [category, filter, colorToFilter]);

  return (
    <Container>
      <header>
        <h2>
          {category === 'calcas'
            ? 'Calças'
            : category === 'calcados'
            ? 'Calçados'
            : category}
        </h2>
        <div>
          {filter && <TagFilter name={filter} onRemove={() => setFilter('')} />}
          {colorToFilter && (
            <TagFilter
              name={colorToFilter}
              onRemove={() => setColorToFilter('')}
            />
          )}
        </div>
      </header>

      <Menu>
        <View>
          <BsGrid3X3GapFill size={20} color="#5dbcd2" />
          <RiLayout2Fill size={24} />
        </View>
        <OrderBy>
          <span>Ordenar por</span>
          <Select
            name="orderBy"
            placeholder="Preço"
            options={orderOptions}
            value="order"
          />
        </OrderBy>
      </Menu>

      <Items>
        {products &&
          products.map((_, index) => (
            <Product
              key={index}
              imageUrl={shirt}
              alt="Tênis azul da Adidas com cadarços rosas"
              name="Tênis Adidas"
              price="R$299,90"
            />
          ))}
      </Items>

      <Paginator />
    </Container>
  );
};

export default Products;
