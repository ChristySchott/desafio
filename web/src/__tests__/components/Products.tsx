import React from 'react';
import { render } from '@testing-library/react';

import Products from '../../components/Products';

describe('<Products />', () => {
  it('should render with no error', () => {
    const { container } = render(<Products />);

    expect(container).toBeTruthy();

    expect(container.firstChild).toMatchSnapshot();
  });
});
