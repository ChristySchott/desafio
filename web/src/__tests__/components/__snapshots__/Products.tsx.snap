// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Products /> should render with no error 1`] = `
<div
  class="sc-kEqYlL hQzvMe"
>
  <header>
    <h2 />
    <div />
  </header>
  <div
    class="sc-iqAbSa fDElvS"
  >
    <div
      class="sc-dIsAE cMOMtg"
    >
      <button
        type="button"
      >
        <svg
          color="#5dbcd2"
          fill="currentColor"
          height="20"
          size="20"
          stroke="currentColor"
          stroke-width="0"
          style="color: rgb(93, 188, 210);"
          viewBox="0 0 16 16"
          width="20"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M1 2a1 1 0 011-1h2a1 1 0 011 1v2a1 1 0 01-1 1H2a1 1 0 01-1-1V2zm5 0a1 1 0 011-1h2a1 1 0 011 1v2a1 1 0 01-1 1H7a1 1 0 01-1-1V2zm5 0a1 1 0 011-1h2a1 1 0 011 1v2a1 1 0 01-1 1h-2a1 1 0 01-1-1V2zM1 7a1 1 0 011-1h2a1 1 0 011 1v2a1 1 0 01-1 1H2a1 1 0 01-1-1V7zm5 0a1 1 0 011-1h2a1 1 0 011 1v2a1 1 0 01-1 1H7a1 1 0 01-1-1V7zm5 0a1 1 0 011-1h2a1 1 0 011 1v2a1 1 0 01-1 1h-2a1 1 0 01-1-1V7zM1 12a1 1 0 011-1h2a1 1 0 011 1v2a1 1 0 01-1 1H2a1 1 0 01-1-1v-2zm5 0a1 1 0 011-1h2a1 1 0 011 1v2a1 1 0 01-1 1H7a1 1 0 01-1-1v-2zm5 0a1 1 0 011-1h2a1 1 0 011 1v2a1 1 0 01-1 1h-2a1 1 0 01-1-1v-2z"
          />
        </svg>
      </button>
      <button
        type="button"
      >
        <svg
          color="#888"
          fill="currentColor"
          height="23.5"
          size="23.5"
          stroke="currentColor"
          stroke-width="0"
          style="color: rgb(136, 136, 136);"
          viewBox="0 0 24 24"
          width="23.5"
          xmlns="http://www.w3.org/2000/svg"
        >
          <g>
            <path
              d="M0 0h24v24H0z"
              fill="none"
            />
            <path
              d="M22 12.999V20a1 1 0 0 1-1 1h-8v-8.001h9zm-11 0V21H3a1 1 0 0 1-1-1v-7.001h9zM11 3v7.999H2V4a1 1 0 0 1 1-1h8zm10 0a1 1 0 0 1 1 1v6.999h-9V3h8z"
            />
          </g>
        </svg>
      </button>
    </div>
    <div
      class="sc-crzoUp kBxbcP"
    >
      <span>
        Ordenar por
      </span>
      <div
        class="sc-jSFkmK dTgRnu"
      >
        <section
          class="sc-gKAblj iMPPEJ"
        >
          <div
            class=" css-2b097c-container"
          >
            <div
              class=" css-yk16xz-control"
            >
              <div
                class=" css-g1d714-ValueContainer"
              >
                <div
                  class=" css-1uccc91-singleValue"
                >
                  Menor preço
                </div>
                <input
                  aria-autocomplete="list"
                  class="css-62g3xt-dummyInput"
                  id="react-select-2-input"
                  readonly=""
                  tabindex="0"
                  value=""
                />
              </div>
              <div
                class=" css-1hb7zxy-IndicatorsContainer"
              >
                <span
                  class=" css-1okebmr-indicatorSeparator"
                />
                <div
                  aria-hidden="true"
                  class=" css-tlfecz-indicatorContainer"
                >
                  <svg
                    aria-hidden="true"
                    class="css-6q0nyr-Svg"
                    focusable="false"
                    height="20"
                    viewBox="0 0 20 20"
                    width="20"
                  >
                    <path
                      d="M4.516 7.548c0.436-0.446 1.043-0.481 1.576 0l3.908 3.747 3.908-3.747c0.533-0.481 1.141-0.446 1.574 0 0.436 0.445 0.408 1.197 0 1.615-0.406 0.418-4.695 4.502-4.695 4.502-0.217 0.223-0.502 0.335-0.787 0.335s-0.57-0.112-0.789-0.335c0 0-4.287-4.084-4.695-4.502s-0.436-1.17 0-1.615z"
                    />
                  </svg>
                </div>
              </div>
            </div>
            <input
              name="orderBy"
              type="hidden"
              value="CRESC"
            />
          </div>
        </section>
      </div>
    </div>
  </div>
  <ul
    class="sc-bqGHjH ewVcGG"
  >
    <div
      class="sc-ksluoS edsndP"
    >
      <svg
        color="#c61a28"
        fill="none"
        height="50"
        size="50"
        stroke="currentColor"
        stroke-linecap="round"
        stroke-linejoin="round"
        stroke-width="2"
        style="color: rgb(198, 26, 40);"
        viewBox="0 0 24 24"
        width="50"
        xmlns="http://www.w3.org/2000/svg"
      >
        <circle
          cx="11"
          cy="11"
          r="8"
        />
        <line
          x1="21"
          x2="16.65"
          y1="21"
          y2="16.65"
        />
      </svg>
      <p>
        Oops! Não encontramos nenhum produto.
      </p>
    </div>
  </ul>
  <section
    class="sc-iCoHVE hINJpc"
  >
    
  </section>
</div>
`;
